File: Codigos estudiantes por lenguaje/Java/Daniel Cangrejo/Ordenamientos/Bitonic_Sort.java
//BITONIC_SORT
public class Ordenamientos2_1 {

	static void compAndSwap(int a[], int i, int j, int dir) {
		
		if ((a[i] > a[j] && dir == 1) || (a[i] < a[j] && dir == 0)) {
			int temp = a[i];
			a[i] = a[j];
			a[j] = temp;
		}
	}
	
	static void bitonicMerge(int a[], int low, int cnt, int dir) {
		if (cnt > 1) {
			int k = cnt/2;
			for (int i = low; i < low +k; i++) {
				compAndSwap(a, i, i+k, dir);
			}
			bitonicMerge(a, low, k, dir);
			bitonicMerge(a, low+k, k, dir);
		}
	}
	
	static void bitonicSort(int a[], int low, int cnt, int dir) {
		if(cnt > 1) {
			int k = cnt/2;
			bitonicMerge(a, low, k, 1);
			bitonicSort(a, low+k, k, 0);
			bitonicMerge(a, low, cnt, dir);
		}
	}
	
	static void sort(int a[],int n, int up) {
		bitonicSort(a, 0, n, up);
	}
	
	static void printArray(int arr[]) {
		int n = arr.length;
		for (int i = 0; i < n; i++) {
			System.out.print(arr[i] + " ");
		}
		System.out.println();
	}
	
	public static void main(String[] args) {
		int a[] = {3,7,4,8,6,2,1,5};
		int up=1;
		//si up es 0 de mayor a menor
		//si es 1 de menor a mayor
		printArray(a);
		sort(a, a.length, up);
		printArray(a);
		

	}

}

File: Codigos estudiantes por lenguaje/PY/Bryann Valderrama/Algoritmos de Ordenamiento/bitonicSort.py
'''Bitonic Sort

   - Hace mas comparaciones que MergeSort.
   - Especializado en lista de doble lazo (Agregar datos por izq y der).
   - Compara los elementos en Secuencia Binotonica.
   - Una secuencia es Bitonica si primero aumenta y luego disminuye.
   - Complejidad Tiempo: O(n log^2 n).
'''

from sys import stdin, stdout
rl = stdin.readline
wr = stdout.write


def compAndSwap(arr, i, j, direc):
    if (arr[i] > arr[j] and direc == 1) or (arr[i] < arr[j] and direc == 0):
        arr[i], arr[j] = arr[j], arr[i]


def bitonicMerge(arr, low, cnt, direc):

    if cnt > 1:
        k = cnt // 2
        for i in range(low, low+k):
            compAndSwap(arr, i, i+k, direc)

        bitonicMerge(arr, low, k, direc)
        bitonicMerge(arr, low+k, k, direc)


def bitonicSort(arr, low, cnt, direc):

    if cnt > 1:
        k = cnt // 2
        bitonicSort(arr, low, k, 1)
        bitonicSort(arr, low+k, k, 0)
        bitonicMerge(arr, low, cnt, direc)


def sortB(arr, n, up):
    bitonicSort(arr, 0, n, up)


def printArray(arr):
    for i in arr:
        wr(f'{i} ')
    wr('\n')


arr = list(map(int, rl().split()))
up = 1  # 0 = Mayor a Menor
printArray(arr)
sortB(arr, len(arr), up)
printArray(arr)
